components/style/values/specified/basic_shape.rs

```diff
@@ -8,12 +8,118 @@
 //! [basic-shape]: https://drafts.csswg.org/css-shapes/#typedef-basic-shape
 
 use cssparser::{Parser, ToCss};
+use parser::{ParserContext, Parse};
 use properties::shorthands::{parse_four_sides, serialize_four_sides};
 use std::fmt;
+use url::Url;
 use values::computed::basic_shape as computed_basic_shape;
 use values::computed::{Context, ToComputedValue, ComputedValueAsSpecified};
 use values::specified::position::Position;
 use values::specified::{BorderRadiusSize, LengthOrPercentage, Percentage};
+use values::specified::UrlExtraData;
+
+/// A shape source, for some reference box
+///
+/// clip-path uses ShapeSource<GeometryBox>,
+/// shape-outside uses ShapeSource<ShapeBox>
+#[derive(Clone, PartialEq, Debug)]
+#[cfg_attr(feature = "servo", derive(HeapSizeOf))]
+pub enum ShapeSource<T> {
+    Url(Url, UrlExtraData),
+    Shape(BasicShape, Option<T>),
+    Box(T),
+    None,
+}
+
+impl<T> Default for ShapeSource<T> {
+    fn default() -> Self {
+        ShapeSource::None
+    }
+}
+
+impl<T: ToCss> ToCss for ShapeSource<T> {
+    fn to_css<W>(&self, dest: &mut W) -> fmt::Result where W: fmt::Write {
+        use values::LocalToCss;
+        match *self {
+            ShapeSource::Url(ref url, _) => url.to_css(dest),
+            ShapeSource::Shape(ref shape, Some(ref reference)) => {
+                try!(shape.to_css(dest));
+                try!(dest.write_str(" "));
+                reference.to_css(dest)
+            }
+            ShapeSource::Shape(ref shape, None) => shape.to_css(dest),
+            ShapeSource::Box(ref reference) => reference.to_css(dest),
+            ShapeSource::None => dest.write_str("none"),
+
+        }
+    }
+}
+
+impl<T: Parse + PartialEq> ShapeSource<T> {
+    pub fn parse(context: &ParserContext, input: &mut Parser) -> Result<Self, ()> {
+        if let Ok(_) = input.try(|input| input.expect_ident_matching("none")) {
+            Ok(ShapeSource::None)
+        } else if let Ok(url) = input.try(|input| input.expect_url()) {
+            match UrlExtraData::make_from(context) {
+                Some(extra_data) => {
+                    Ok(ShapeSource::Url(context.parse_url(&url), extra_data))
+                },
+                None => Err(()),
+            }
+        } else {
+            let mut source = ShapeSource::None;
```

Comment:
What do you think about making this block a little more concise?  Something like this (untested):

```rust
fn parse_component<U>(input: &mut Parser, component: &mut Option<U>) -> bool {
    if component.is_some() {
        return false;  // already parsed this component
    }
    *component = input.try(U::parse).ok();
    component.is_some()
}

let mut shape = None;
let mut reference = None;
loop {
    if !Self.parse_component(input, shape) &&
       !Self.parse_component(input, reference) {
        break;
    }
}
match (shape, reference) => {
    (Some(shape), _) => Ok(ShapeSource::Shape(shape, reference)),
    (None, Some(reference)) => Ok(ShapeSource::Box(reference)),
    (None, None) => Err(()),
}
```